adal/src/main/java/com/microsoft/aad/adal/AcquireTokenRequest.java

```diff
@@ -168,11 +169,58 @@ public void run() {
+
+                    final String correlationId = String.format(" CorrelationId: %s", authenticationRequest.getCorrelationId().toString());
+
+                    if (result == null) {
+                        Logger.e(TAG + methodName, "Returned result with exchanging refresh token for access token is null" + correlationId, "",
+                                ADALError.AUTH_REFRESH_FAILED);
+                        throw new AuthenticationException(
+                                ADALError.AUTH_REFRESH_FAILED, correlationId);
+                    }
+
+                    if (!StringExtensions.isNullOrBlank(result.getErrorCode())) {
+                        Logger.e(TAG + methodName, " ErrorCode:" + result.getErrorCode(), " ErrorDescription:" + result.getErrorDescription(), ADALError.AUTH_REFRESH_FAILED);
+                        throw new AuthenticationException(ADALError.AUTH_REFRESH_FAILED,
+                                " ErrorCode:" + result.getErrorCode());
+                    }
+
+                    if (StringExtensions.isNullOrBlank(result.getAccessToken())) {
+                        Logger.e(TAG + methodName, "Access Token not returned from server", "", ADALError.AUTH_FAILED_NO_TOKEN);
+                        throw new AuthenticationException(ADALError.AUTH_FAILED_NO_TOKEN,
+                                " Access Token not returned from server ");
+                    }
+
+                    final String rawIdToken = result.getIdToken();
+
+                    if (!TextUtils.isEmpty(rawIdToken)) {
+                        final IdToken idTokenRecord = new IdToken(rawIdToken);
+                        final UserInfo userInfo = new UserInfo(idTokenRecord);
+
+                        TokenCacheItem tokenCacheItem = new TokenCacheItem();
+                        tokenCacheItem.setRawIdToken(rawIdToken);
+                        tokenCacheItem.setUserInfo(userInfo);
+                        tokenCacheItem.setTenantId(idTokenRecord.getTenantId());
+
+                        // This method is used in ADAL to store tokens in cache during a silent aka
+                        // refresh token request. We are going to be using the same method for this
+                        // API as well for consistency.
+                        //
+                        // To get this to work, we created a TokenCacheItem object above and
+                        // populate it with an ID Token and UserInfo object as in this API ADAL does
+                        // not do a cache look up and does not have the ID Token in cache already.
+                        // We got an ID Token from eSTS and will be using that ID Token to create
+                        // a UserInfo object and seed it into the cache along with the other tokens.
+                        // This allows us to achieve SSO for this API.
+                        mTokenCacheAccessor.updateCachedItemWithResult(authenticationRequest,
```

Comment:
Slightly confused by this; the javadoc for this method describes the cache as unused, but now we're updating the cache

```java
    /**
     * Developer is using refresh token call to do refresh without cache usage.
     * App context or activity is not needed. Async requests are created, so this
     * needs to be called at UI thread.
     */
```

Reply:
The repercussions of this change aren't entirely clear to me. Previously this method was used in the `acquireTokenByRefreshToken` flow where IIRC _nothing_ was cached. Now that caching is being added, won't this change the flow of the code thru ADAL for consumers of this API upon subsequent use due to the newly cached RT now being saved in the cache? Is there any scenario we may not be considering whereby customers may be _relying_ on ADAL not caching this token?

Reply:
Synced offline; this change should be fine. But docs should be updated

Reply:
Addressed. Updated docs.